source/socket.c

```diff
@@ -0,0 +1,267 @@
+}
+
+int aws_socket_assign_to_event_loop(struct aws_socket *socket, struct aws_event_loop *event_loop) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_assign_to_event_loop_fn);
+    return socket->vtable->socket_assign_to_event_loop_fn(socket, event_loop);
+}
+
+struct aws_event_loop *aws_socket_get_event_loop(struct aws_socket *socket) {
+    return socket->event_loop;
+}
+
+int aws_socket_subscribe_to_readable_events(
+    struct aws_socket *socket,
+    aws_socket_on_readable_fn *on_readable,
+    void *user_data) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_subscribe_to_readable_events_fn);
+    return socket->vtable->socket_subscribe_to_readable_events_fn(socket, on_readable, user_data);
+}
+
+int aws_socket_read(struct aws_socket *socket, struct aws_byte_buf *buffer, size_t *amount_read) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_read_fn);
+    return socket->vtable->socket_read_fn(socket, buffer, amount_read);
+}
+
+int aws_socket_write(
+    struct aws_socket *socket,
+    const struct aws_byte_cursor *cursor,
+    aws_socket_on_write_completed_fn *written_fn,
+    void *user_data) {
+
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_write_fn);
+    return socket->vtable->socket_write_fn(socket, cursor, written_fn, user_data);
+}
+
+int aws_socket_get_error(struct aws_socket *socket) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_get_error_fn);
+    return socket->vtable->socket_get_error_fn(socket);
+}
+
+bool aws_socket_is_open(struct aws_socket *socket) {
+    AWS_PRECONDITION(socket->vtable && socket->vtable->socket_is_open_fn);
+    return socket->vtable->socket_is_open_fn(socket);
+}
+
+static enum aws_socket_impl_type aws_socket_get_default_impl_type(void);
+static int aws_socket_impl_type_validate_platform(enum aws_socket_impl_type type);
+int aws_socket_init(struct aws_socket *socket, struct aws_allocator *alloc, const struct aws_socket_options *options) {
+
+    // 1. get socket type & validate type is avliable the platform
+    enum aws_socket_impl_type type = options->impl_type;
+    if (type == AWS_SIT_PLATFORM_DEFAULT) {
+        type = aws_socket_get_default_impl_type();
+    }
+
+    if (aws_socket_impl_type_validate_platform(type)) {
+        AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "Invalid event loop type on the platform.");
+        return AWS_ERROR_PLATFORM_NOT_SUPPORTED;
+    }
+
+    // 2. setup vtable based on socket type
+    switch (type) {
+        case AWS_SIT_POSIX:
+            return aws_socket_init_posix(socket, alloc, options);
+            break;
+        case AWS_SIT_WINSOCK:
+            return aws_socket_init_winsock(socket, alloc, options);
+            break;
+
+        case AWS_SIT_APPLE_NETWORK_FRAMEWORK:
+            AWS_ASSERT(false && "Invalid socket implementation on platform.");
+            return aws_socket_init_apple_nw_socket(socket, alloc, options);
```

Comment:
why is this assert(false) here?

Reply:
At the point of this PR, apple network framework is not supported yet. Added the assertion to warn developer that we should never use this type so far. 
I will update the comments to explan it.